### ISSUE TITLE
JupyterNotebook Ram Memory + data validating question

### ISSUE THREAD

**Issue by armanivers** [ultralytics/ultralytics#18724]
I'm using a pre-trained segmentation model on a custom dataset. My goal is to verify the performance of the model.

**First question:**
Does model.val() use the best.pt weights from the training folder? Or should i load the best.pt into a model and do a model.val() off of it?
(Example 1)

**Second question:**
When the model is training, does it use the val data or does it JUST focus on the train data. Is the way I'm splitting the train+val data into kfolds and validating model 1 (trained on kfold1) on val-data 1 correct?

**Final Question/Problem:**
I run the clear memory and delete all the variables I've created at the end of every training. My goal is to clear all the memory from training.
The problem that I face (I run the script in JupyterNotebooks but also tried in just python cmd in linux) after the first training is finished, the ram memory is not refreshed. It only goes to zero, if I manually reset the kernel for the notebook. Is there a better way to dispose of the whole training data before starting new training? **After around 10 kfolds (each 100 epochs) the script stops running.**

**My folder structure looks like this:**

Root
->train (full dataset - including ground truth)
->val (ful dataset - including ground truth)
->kfold
--->kfold1
------>train
------>val
--->kfold2
------>train
------>val
--->kfold3
------>train
------>val

**Config:**
config1.yaml
names:
  0: object
path: root/kfold/kfold1
train: train/images
val: val/images

config2.yaml
names:
  0: object
path: root/kfold/kfold2
train: train/images
val: val/images

**Code:**

```
def clear_memory():
    gc.collect()
    if torch.cuda.is_available():
        torch.cuda.empty_cache()
        torch.cuda.reset_peak_memory_stats()
        torch.cuda.synchronize()

#load pretrained model
model = YOLO('yolov8x-seg.pt')
------------------------------------------------------------------------
#train model on kfold1
model_results = model.train(data=os.path.join(ROOT_DIR, 'config1.yaml'), imgsz=512, batch=16, deterministic=True, plots=True,
                    close_mosaic = 0,
                    optimizer = "SGD", 
                    epochs = 120, 
                    # disable built-in augmentation, instead use Albumentations Library
                    augment=False, hsv_h=0, hsv_s=0, hsv_v=0, degrees=0, translate=0, 
                    scale=0, shear=0.0, perspective=0, flipud=0, fliplr=0, bgr=0, 
                    mosaic=0, mixup=0, copy_paste=0, erasing=0, crop_fraction=0)

**#Example 1**
#evaluate model's performance on validation dataset
val_results= model.val()
print(val_results)

#should be the same as above right?
model = YOLO("path/to/trainingfolder/weights/best.pt")
val_results_best = model.val() 
------------------------------------------------------------------------
del model
del model_results
del val_results
clear_memory()
#Training of kfold 2
------------------------------------------------------------------------
#load pretrained model
model = YOLO('yolov8x-seg.pt')
#train model on kfold1
model_results = model.train(data=os.path.join(ROOT_DIR, 'config1.yaml'), imgsz=512, batch=16, deterministic=True, plots=True,
                    close_mosaic = 0,
                    optimizer = "SGD", 
                    epochs = 120, 
                    # disable built-in augmentation, instead use Albumentations Library
                    augment=False, hsv_h=0, hsv_s=0, hsv_v=0, degrees=0, translate=0, 
                    scale=0, shear=0.0, perspective=0, flipud=0, fliplr=0, bgr=0, 
                    mosaic=0, mixup=0, copy_paste=0, erasing=0, crop_fraction=0)

**#Example 1**
#evaluate model's performance on validation dataset
val_results= model.val()
print(val_results)

#should be the same as above right?
model = YOLO("path/to/trainingfolder/weights/best.pt")
val_results_best = model.val() 
------------------------------------------------------------------------
del model
del model_results
del val_results
clear_memory()
#Training of kfold 3...
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @armanivers, thank you for your detailed issue report and your interest in Ultralytics 🚀! We appreciate the thorough context you've provided regarding your usage scenario and the challenges you're encountering.

For your questions and concerns, here are some next steps:

### 🐛 Bug Reporting
If you believe the issue with memory not clearing between trainings and requiring a kernel reset is a bug, please help us debug it by providing a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/). This will allow us to pinpoint the issue more effectively. For example, you could strip down your script to the smallest functional code that reproduces the memory behavior.

### ❓ Questions and Best Practices
For your training and validation questions:
- Ensure you’re following the guidance in the [model training tips](https://docs.ultralytics.com/guides/model-training-tips/) for best results.
- Additional details such as dataset structure, training logs, and the behavior you're expecting versus observing can help clarify your queries.

### Memory Management
For issues with memory persistence after training:
- Consider ensuring your `clear_memory()` function is sufficient by testing it in smaller cases.
- If you're encountering repeated GPU memory build-up, sharing your PyTorch and CUDA versions, as well as the exact platform and configurations you’re using (libraries, drivers, etc.), would be helpful for deeper investigation.

### Upgrade
It’s always a good idea to verify your issue on the latest version of Ultralytics and its dependencies. Upgrade with:

```bash
pip install -U ultralytics
```

This ensures any bug fixes or improvements made in recent versions are applied.

### Community and Support
If you'd like to discuss your approach further or share your experience with other users, feel free to join the Ultralytics community:
- [Discord](https://discord.com/invite/ultralytics) for real-time help and discussions 🎧
- [Discourse](https://community.ultralytics.com) for in-depth chat on methodology and workflows
- [Subreddit](https://reddit.com/r/Ultralytics) to exchange knowledge and ideas with others in the community

### Verified Environments
If applicable, try running your experiments in one of our verified environments, which have pre-installed dependencies, including [CUDA](https://developer.nvidia.com/cuda), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/). Examples include:
- <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a>   
- <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>  
- <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

### Status
Our CI pipeline confirms the correctness of YOLO across tasks, environments, and platforms regularly. You can verify its current status here:

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

This is an automated response to assist you quickly, but rest assured that an Ultralytics engineer will review your issue in detail soon. Stay tuned! 😊

**Comment by Y-T-G** [comment]
> Does model.val() use the best.pt weights from the training folder? Or should i load the best.pt into a model and do a model.val() off of it?

You should load it. And there's no point in doing `model.val()` right after `model.train()` because training automatically does `model.val()` after it ends.

> When the model is training, does it use the val data or does it JUST focus on the train data. Is the way I'm splitting the train+val data into kfolds and validating model 1 (trained on kfold1) on val-data 1 correct?

No deep learning training uses training data for evaluation. It defeats the point of evaluation.

> The problem that I face (I run the script in JupyterNotebooks but also tried in just python cmd in linux) after the first training is finished, the ram memory is not refreshed. It only goes to zero, if I manually reset the kernel for the notebook. Is there a better way to dispose of the whole training data before starting new training?

You don't need to clear memory manually. Python automatically does that as needed. It's pointless.

**Comment by armanivers** [comment]
1.
So it's correct for me to load the best.pt of the training and then use model.val() off of the loaded model using the validation dataset of the kfold? (after the training has finished)

2. Yes you're right, I was a bit confused whether the val is used actively in the training for the training or just as a way to tell the model how it performs, sorry

3. So that leaves the problem of Jupyternotebook not being able to handle many yolov8 trainin processes after one another? Would it be smarter to use the CLI for that then instead? Or what is your general approach when you run around 10 trainings with each around 100 epochs, do you not have such problems?

**Comment by armanivers** [comment]
also another option I saw was:

model = YOLO("yolov8x-seg") # load pretrained model
model.load("path/to/train/weighets/best.pt") #apply weights from trained model
model.val() #validate the final model

but that is wrong right? I don't need to load the best weights into the pre-trained model, since the best.pt already includes everything needed.

**Comment by Y-T-G** [comment]
1. Load it and run val.
3. Python automatically removes allocated resources for objects that are no longer being used. If you want a fresh environment, CLI creates a fresh Python process each time.
